### PR TITLE
Deprecate Python 3.7

### PR DESCRIPTION
**Pull request by adamjstewart**
As of yesterday, Python 3.7 has officially reached [EOL](https://endoflife.date/python). This PR deprecates Python 3.7 installation in Spack. Python 3.7 support will be removed after the next minor release.

FYI @manuelakuhn @alalazo @tgamblin

### CODE REVIEW

**Comment by adamjstewart** [comment]
Pipelines are failing because some of the packages in the data-vis-sdk pipeline rely on Python 3.7. Not entirely sure why but some package is requiring VTK 8 which doesn't support Python 3.8+. Is there a way to use newer VTK in the pipeline by adding a new version or relaxing a constraint somewhere? Pinging maintainers of some of these packages: @kwryankrattiger @svenevs @chuckatkins @danlipsa @cyrush @EthanS94

**Comment by kwryankrattiger** [comment]
TL;DR, that dependency isn't going away anytime soon.

That python requirement comes form VisIt. The discussion about this is here https://github.com/visit-dav/visit/discussions/18662 . While that will address future versions of VisIt, the not so old versions released as recently as Feb 2023 should probably not get dropped yet.

Not that I personally use the older Python v3 versions, I do wonder what the consequences of dropping those have on some packages that need older APIs that are not deprecated and not tested in CI.

I get wanting to drop old packages, but I don't know if this one is quite ready. I think this may also need additional consideration as pointed out for this PR https://github.com/spack/spack/pull/37895 in [slack](https://spackpm.slack.com/archives/C03LVJYGZ7X/p1686782262859889).

**Comment by tgamblin** [comment]
@kwryankrattiger: there are apparently (according to @mathstuf) fixes for VTK 8.2 to use Python 3.8 here: https://discourse.vtk.org/t/python-3-8-wrapping-support/2621/2.

Can you either throw those onto VTK8 as a patch in Spack, or cut an 8.2.1 release that would include them?

**Comment by kwryankrattiger** [comment]
> Can you either throw those onto VTK8 as a patch in Spack, or cut an 8.2.1 release that would include them?

I don't think an 8.2.1 is going to happen considering it is pretty old at this point and most other projects have moved to VTK 9 at this point for their development/releases.

I am not opposed to patching VTK 8, but there are a number of other VTK releases requiring `depends_on("python@:3.7", when="@:8", type=("build", "run"))` which are not deprecated. So removing python 3.7 is still going to break those and I doubt very much that those patches apply outside of VTK 8, or even outside of VTK 8.2.0, but we could test that if that is the way we want to go.

I am kind of surprised that the `audit-packages` CI checks don't complain about this since there are roughly 489 packages that reference python 3.7 (just a quick grep `grep -R "python@.*3.7" | grep -v Binary | wc -l`). So regardless of fixing up VTK to build with python 3.8, I don't think that is solving the problems of removing python 3.7.

**Comment by kwryankrattiger** [comment]
A faster fix is probably to just turn off VisIt in the DaV SDK pipeline. I left it there as an optimistic test, but officially the Data and Vis SDK does not support VisIt due to the python issue.

https://ecp-data-vis-sdk.github.io/#visit

https://github.com/spack/spack/pull/38642

**Comment by adamjstewart** [comment]
> I am kind of surprised that the audit-packages CI checks don't complain about this

The versions still exist, `spack audit packages` only complains when a version doesn't exist, not when it's deprecated. Once this PR is merged, users can still install VTK 8, but Spack will ask them if they are sure.

> there are roughly 489 packages that reference python 3.7

The vast majority of those mention it as a lower bound, not as an upper bound. The ones that do list it as an upper bound are:
```console
$ spack pkg grep 'depends_on..python@.*:3.7'
/Users/Adam/spack/var/spack/repos/builtin/packages/catalyst/package.py:    depends_on("python@3:3.7", when="@:5.7 +python", type=("build", "run"))
/Users/Adam/spack/var/spack/repos/builtin/packages/gaudi/package.py:    depends_on("python@:3.7", when="@32.2:34", type=("build", "run"))
/Users/Adam/spack/var/spack/repos/builtin/packages/libpeas/package.py:    depends_on("python@3:3.7.9", type="build")
/Users/Adam/spack/var/spack/repos/builtin/packages/node-js/package.py:    depends_on("python@2.7,3.5:3.7", when="@12:13.0", type="build")
/Users/Adam/spack/var/spack/repos/builtin/packages/paraview/package.py:    depends_on("python@3:3.7", when="@:5.7 +python", type=("build", "run"))
/Users/Adam/spack/var/spack/repos/builtin/packages/py-azureml-dataprep-native/package.py:    depends_on("python@3.7.0:3.7", when="@30.0.0-py3.7", type=("build", "run"))
/Users/Adam/spack/var/spack/repos/builtin/packages/py-azureml-dataprep-native/package.py:    depends_on("python@3.7.0:3.7", when="@14.2.1-py3.7", type=("build", "run"))
/Users/Adam/spack/var/spack/repos/builtin/packages/py-azureml-dataprep-rslex/package.py:    depends_on("python@3.7.0:3.7", when="@1.9.0-py3.7,1.8.0-py3.7", type=("build", "run"))
/Users/Adam/spack/var/spack/repos/builtin/packages/py-cvxopt/package.py:    depends_on("python@2.7:3.7", type=("build", "link", "run"), when="@:1.1.9")
/Users/Adam/spack/var/spack/repos/builtin/packages/py-itk/package.py:    depends_on("python@3.7.0:3.7", when="@5.1.1-cp37,5.1.2-cp37,5.3.0-cp37", type=("build", "run"))
/Users/Adam/spack/var/spack/repos/builtin/packages/py-numba/package.py:    depends_on("python@3.3:3.7", when="@0.40.1:0.47", type=("build", "run"))
/Users/Adam/spack/var/spack/repos/builtin/packages/py-numpy/package.py:    depends_on("python@:3.7", when="@1.14.5:1.17.2", type=("build", "link", "run"))
/Users/Adam/spack/var/spack/repos/builtin/packages/py-pillow/package.py:    depends_on("python@2.7:2.8,3.5:3.7", when="@6.0:6.2.0", type=("build", "run"))
/Users/Adam/spack/var/spack/repos/builtin/packages/py-pillow/package.py:    depends_on("python@2.7:2.8,3.4:3.7", when="@5.2:5.4", type=("build", "run"))
/Users/Adam/spack/var/spack/repos/builtin/packages/py-psycopg2/package.py:    depends_on("python@:3.7", when="@:2.8.3", type=("build", "link", "run"))
/Users/Adam/spack/var/spack/repos/builtin/packages/py-pycairo/package.py:    depends_on("python@2.7:2.8,3.4:3.7", when="@1.18.1:1.19", type=("build", "run"))
/Users/Adam/spack/var/spack/repos/builtin/packages/py-pygit2/package.py:    depends_on("python@2.7:3.7", when="@0.28:0")
/Users/Adam/spack/var/spack/repos/builtin/packages/py-pyspark/package.py:    depends_on("python@2.7:3.7", when="@:2", type=("build", "run"))
/Users/Adam/spack/var/spack/repos/builtin/packages/py-quantities/package.py:    depends_on("python@2.7.0:2.7,3.4:3.7", type=("build", "run"), when="@0.12.3")
/Users/Adam/spack/var/spack/repos/builtin/packages/py-rasterio/package.py:    depends_on("python@2.7:2.8,3.5:3.7", when="@:1.0", type=("build", "link", "run"))
/Users/Adam/spack/var/spack/repos/builtin/packages/py-scipy/package.py:    depends_on("python@:3.7", when="@1.1:1.3.1", type=("build", "link", "run"))
/Users/Adam/spack/var/spack/repos/builtin/packages/py-shared-memory38/package.py:    depends_on("python@3.6:3.7", type=("build", "run"))
/Users/Adam/spack/var/spack/repos/builtin/packages/py-tap-py/package.py:    depends_on("python@3.5:3.7", when="@3.0:", type=("build", "run"))
/Users/Adam/spack/var/spack/repos/builtin/packages/py-tap-py/package.py:    depends_on("python@2.7:2.8,3.5:3.7", type=("build", "run"))
/Users/Adam/spack/var/spack/repos/builtin/packages/py-tensorflow/package.py:    depends_on("python@2.7,3.5:3.7", when="@:2.1", type=("build", "run"))
/Users/Adam/spack/var/spack/repos/builtin/packages/py-torch/package.py:    depends_on("python@2.7:2,3.5:3.7", when="@:1.3", type=("build", "link", "run"))
/Users/Adam/spack/var/spack/repos/builtin/packages/py-torchvision/package.py:    depends_on("python@2.7,3.5:3.7", when="@:0.4", type=("build", "link", "run"))
/Users/Adam/spack/var/spack/repos/builtin/packages/py-x21/package.py:    depends_on("python@3.7.0:3.7", type=("build", "run"), when="@0.2.6-py3.7")
/Users/Adam/spack/var/spack/repos/builtin/packages/visit/package.py:    depends_on("python@3.2:3.7,3.9:", when="+python")
/Users/Adam/spack/var/spack/repos/builtin/packages/vtk/package.py:        depends_on("python@:3.7", when="@:8", type=("build", "run"))
```
Most of these are old versions, not all versions.

**Comment by cyrush** [comment]
I guess this means VisIt will no longer be tested in Spack CI?

**Comment by adamjstewart** [comment]
Not only that, but VTK 8 and older will be removed from Spack completely unless they are patched to support modern Python versions. We won't be removing Python 3.7 until after the next minor release (November) so there's plenty of time to do this.

**Comment by kwryankrattiger** [comment]
> I guess this means VisIt will no longer be tested in Spack CI?

@cyrush 
Correct. I tried to keep it in there as long as possible because I think it is important to continue to test that package. I am hopeful that the VisIt VTK 9 update lands soon in a release, at that time I would turn it back on in the PR that adds the newer version. I left the stack config there so it should be pretty easy to do.

**Comment by tgamblin** [comment]
I talked to @kwryankrattiger about this -- he is going to add a patch to vtk 8.2.0 so that we can keep VisIt on in the pipelines.  Not keeping VisIt on in Spack CI is going to cause headaches for them, as VisIt is actively used on Perlmutter and Frontier via Spack.  I don't want that to break.

VTK currently has the following constraints:

```python
    with when("+python"):
        # Depend on any Python, add bounds below.
        extends("python@2.7:", type=("build", "run"))
        # Python 3.8 support from vtk 9
        depends_on("python@:3.7", when="@:8", type=("build", "run"))
        # Python 3.10 support from vtk 9.2
        depends_on("python@:3.9", when="@:9.1", type=("build", "run"))
```

I think it's easy enough to patch the 8.2.0 version and change the middle constraint to:

```python
        # Python 3.8 support from vtk 9 (and 8.2.0 with post-release patches)
        depends_on("python@:3.7", when="@:8.1", type=("build", "run"))
```

**Review by manuelakuhn — APPROVED** [review]
I don't have a strong opinion about this. LGTM

**Pull request merged**
